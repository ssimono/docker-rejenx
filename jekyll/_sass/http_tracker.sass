.http-tracker
  select[name="method"]
    padding-bottom: 6px

  .api-root
    text-align: center
    margin: 0 2px
