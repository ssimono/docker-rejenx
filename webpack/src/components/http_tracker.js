import React, { Component } from 'react';
import PropTypes from 'prop-types';

function renderCall(netCall, idx) {
  let color = 'grey';
  if (netCall.error) {
    color = 'purple';
  } else if (netCall.res) {
    color = netCall.res.ok ? 'green' : 'red';
  }

  return (
    <tr key={(v => v)(idx)}>
      <td>
        <i className={`fa fa-circle ${color}`} aria-hidden="true" />&nbsp;
        {netCall.res && `${netCall.res.status}`}
        {netCall.error && `${netCall.error.message}`}
      </td>
      <td><strong>{netCall.req.method}</strong> {netCall.req.url}</td>
      <td><code>{netCall.body}</code></td>
    </tr>
  );
}

// component
export function HttpTracker(props) {
  return (
    <div className="http-tracker">
      <h3>Test access to the API</h3>
      <form onSubmit={props.onStartCall} className="ink-form column-group">
        <div className="control-group">
          <select
            name="method"
            onChange={props.onEditRequest}
            value={props.editedRequest.method}
          >

            <option value="GET">GET</option>
            <option value="POST">POST</option>
          </select>
        </div>

        <div className="control-group api-root">
          <mark>/</mark>
        </div>

        <div className="control-group">
          <div className="control append-button">
            <span>
              <input
                name="url"
                onChange={props.onEditRequest}
                type="text"
                value={props.editedRequest.url}
              />
            </span>
            <input type="submit" value="OK" className="ink-button" />
          </div>
        </div>
      </form>
      <table className="ink-table alternating">
        <thead>
          <tr>
            <td>Status</td>
            <td>Request</td>
            <td>Body</td>
          </tr>
        </thead>
        <tbody>{ props.calls.map(renderCall) }</tbody>
      </table>
    </div>
  );
}

// Container
export default class HttpTrackerContainer extends Component {
  constructor(props) {
    super(props);
    this.state = {
      calls: [],
      editedRequest: { method: 'GET', url: '' },
    };

    this.handleEditRequest = this.handleEditRequest.bind(this);
    this.handleStartRequest = this.handleStartRequest.bind(this);
  }

  handleEditRequest(evt) {
    switch (evt.target.name) {
      case 'url':
      case 'method':
        this.setState({ editedRequest: Object.assign(
          {},
          this.state.editedRequest,
          { [evt.target.name]: evt.target.value },
        ) });
        break;
      default:
        break;
    }
  }

  handleStartRequest(evt) {
    const currentIndex = this.state.calls.length;
    const HttpRequest = new Request(
      `${this.props.api_url}/${this.state.editedRequest.url}`,
      { method: this.state.editedRequest.method },
    );

    evt.preventDefault();
    this.setState({
      calls: [].concat(this.state.calls, { req: HttpRequest, res: null }),
    });

    fetch(HttpRequest).then((response) => {
      const calls = [].concat(this.state.calls);

      response.text().then((body) => {
        calls[currentIndex].res = response;
        calls[currentIndex].body = body;
        this.setState({ calls });
      });
    }).catch((err) => {
      const calls = [].concat(this.state.calls);
      calls[currentIndex].error = err;
      this.setState({ calls });
    });
  }

  render() {
    return (
      <HttpTracker
        calls={this.state.calls}
        editedRequest={this.state.editedRequest}
        onEditRequest={this.handleEditRequest}
        onStartCall={this.handleStartRequest}
      />
    );
  }
}

HttpTracker.propTypes = {
  onStartCall: PropTypes.func.isRequired,
  onEditRequest: PropTypes.func.isRequired,
  editedRequest: PropTypes.shape({
    method: PropTypes.oneOf(['GET', 'POST', 'PUT', 'PATCH', 'DELETE', 'OPTION']).isRequired,
    url: PropTypes.string,
  }).isRequired,
  calls: PropTypes.arrayOf(PropTypes.shape({
    req: PropTypes.instanceOf(global.Request),
    res: PropTypes.instanceOf(global.Response),
    err: PropTypes.instanceOf(Error),
  })).isRequired,
};

HttpTrackerContainer.propTypes = {
  api_url: PropTypes.string.isRequired,
};
