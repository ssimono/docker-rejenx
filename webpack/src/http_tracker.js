import React, { Component } from 'react';
import PropTypes from 'prop-types';

function renderRequest(request, idx) {
  let color = 'grey';
  if (request.error) {
    color = 'purple';
  } else if (request.res) {
    color = request.res.ok ? 'green' : 'red';
  }

  return (
    <tr key={(v => v)(idx)}>
      <td>
        <i className={`fa fa-circle ${color}`} aria-hidden="true" />&nbsp;
        {request.res && `${request.res.status}`}
        {request.error && `${request.error.message}`}
      </td>
      <td><strong>{request.req.method}</strong> {request.req.url}</td>
      <td><code>{request.body}</code></td>
    </tr>
  );
}

export default class HttpTracker extends Component {
  constructor(props) {
    super(props);
    this.state = {
      requests: [],
      count: null,
      currentRequest: { method: 'GET', url: '' },
    };

    this.handleEditRequest = this.handleEditRequest.bind(this);
    this.handleStartRequest = this.handleStartRequest.bind(this);
  }

  handleEditRequest(evt) {
    switch (evt.target.name) {
      case 'url':
      case 'method':
        this.setState({ currentRequest: Object.assign(
          {},
          this.state.currentRequest,
          { [evt.target.name]: evt.target.value },
        ) });
        break;
      default:
        break;
    }
  }

  handleStartRequest(evt) {
    const currentIndex = this.state.requests.length;
    const HttpRequest = new Request(
      `${this.props.api_url}/${this.state.currentRequest.url}`,
      { method: this.state.currentRequest.method },
    );

    evt.preventDefault();
    this.setState({
      requests: [].concat(this.state.requests, { req: HttpRequest, res: null }),
    });

    fetch(HttpRequest).then((response) => {
      const requests = [].concat(this.state.requests);

      response.text().then((body) => {
        requests[currentIndex].res = response;
        requests[currentIndex].body = body;
        this.setState({ requests });
      });
    }).catch((err) => {
      const requests = [].concat(this.state.requests);
      requests[currentIndex].error = err;
      this.setState({ requests });
    });
  }

  render() {
    return (
      <div className="http-tracker">
        <h3>Test access to the API</h3>
        <form onSubmit={this.handleStartRequest} className="ink-form column-group">
          <div className="control-group">
            <select
              name="method"
              onChange={this.handleEditRequest}
              value={this.state.currentRequest.method}
            >

              <option value="GET">GET</option>
              <option value="POST">POST</option>
            </select>
          </div>

          <div className="control-group api-root">
            <mark>/</mark>
          </div>

          <div className="control-group">
            <div className="control append-button">
              <span>
                <input
                  name="url"
                  onChange={this.handleEditRequest}
                  type="text"
                  value={this.state.currentRequest.url}
                />
              </span>
              <input type="submit" value="OK" className="ink-button" />
            </div>
          </div>
        </form>
        <table className="ink-table alternating">
          <thead>
            <tr>
              <td>Status</td>
              <td>Request</td>
              <td>Body</td>
            </tr>
          </thead>
          <tbody>{ this.state.requests.map(renderRequest) }</tbody>
        </table>

      </div>
    );
  }
}

HttpTracker.propTypes = {
  api_url: PropTypes.string.isRequired,
};
