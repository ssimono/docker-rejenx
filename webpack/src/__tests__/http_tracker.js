import React from 'react';
import renderer from 'react-test-renderer';
import { HttpTracker } from '../components/http_tracker';

it('renders as expected', () => {
  const mockStart = jest.fn();
  const mockEdit = jest.fn();
  const editedRequest = { method: 'POST', url: '' };

  const call1 = {
    req: new Request('https://www.google.com/lol'),
    res: null,
  };

  const call2 = {
    req: new Request('https://www.google.com/lol', { method: 'POST' }),
    res: new Response(),
  };

  const call3 = {
    req: new Request('https://www.google.com/lol'),
    error: new Error('Something went wrong!!'),
  };

  const tree = renderer.create(
    <HttpTracker
      onStartCall={mockStart}
      onEditRequest={mockEdit}
      editedRequest={editedRequest}
      calls={[call1, call2, call3]}
    />,
  ).toJSON();

  expect(tree).toMatchSnapshot();
});
