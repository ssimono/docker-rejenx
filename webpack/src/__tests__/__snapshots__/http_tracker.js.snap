// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders as expected 1`] = `
<div
  className="http-tracker"
>
  <h3>
    Test access to the API
  </h3>
  <form
    className="ink-form column-group"
    onSubmit={[Function]}
  >
    <div
      className="control-group"
    >
      <select
        name="method"
        onChange={[Function]}
        value="POST"
      >
        <option
          value="GET"
        >
          GET
        </option>
        <option
          value="POST"
        >
          POST
        </option>
      </select>
    </div>
    <div
      className="control-group api-root"
    >
      <mark>
        /
      </mark>
    </div>
    <div
      className="control-group"
    >
      <div
        className="control append-button"
      >
        <span>
          <input
            name="url"
            onChange={[Function]}
            type="text"
            value=""
          />
        </span>
        <input
          className="ink-button"
          type="submit"
          value="OK"
        />
      </div>
    </div>
  </form>
  <table
    className="ink-table alternating"
  >
    <thead>
      <tr>
        <td>
          Status
        </td>
        <td>
          Request
        </td>
        <td>
          Body
        </td>
      </tr>
    </thead>
    <tbody>
      <tr>
        <td>
          <i
            aria-hidden="true"
            className="fa fa-circle grey"
          />
           
        </td>
        <td>
          <strong>
            GET
          </strong>
           
          https://www.google.com/lol
        </td>
        <td>
          <code />
        </td>
      </tr>
      <tr>
        <td>
          <i
            aria-hidden="true"
            className="fa fa-circle red"
          />
           
          undefined
        </td>
        <td>
          <strong>
            POST
          </strong>
           
          https://www.google.com/lol
        </td>
        <td>
          <code />
        </td>
      </tr>
      <tr>
        <td>
          <i
            aria-hidden="true"
            className="fa fa-circle purple"
          />
           
          Something went wrong!!
        </td>
        <td>
          <strong>
            GET
          </strong>
           
          https://www.google.com/lol
        </td>
        <td>
          <code />
        </td>
      </tr>
    </tbody>
  </table>
</div>
`;
